### PR TITLE
improve remotecommand testing fuzzing the data stream

### PR DESCRIPTION
**Pull request by aojea**
/kind cleanup
```release-note
NONE
```

I couldn't understand the behavior observer in the e2e library that uses the remote_command code
https://github.com/kubernetes/kubernetes/pull/115323#issuecomment-1444700603

so I created a test to prove myself we don't have something wrong.

/cc @pohly 
/assign @liggitt @lavalamp 
/sig api-machinery

```
stress ./remotecommand.test -test.run TestStreamRandomData
5s: 1203 runs so far, 0 failures
10s: 2451 runs so far, 0 failures
15s: 3716 runs so far, 0 failures
20s: 4975 runs so far, 0 failures
25s: 6235 runs so far, 0 failures
30s: 7489 runs so far, 0 failures
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go [hunk 1]
```diff
@@ -365,3 +368,74 @@ func TestStreamExitsAfterConnectionIsClosed(t *testing.T) {
 		return
 	}
 }
+
+func TestStreamRandomData(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var stdin, stdout bytes.Buffer
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: Can you use https://pkg.go.dev/io#Pipe instead?

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: heh, I started with that approach  but the race detector complains , so I need to create this buffers all over the test

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: That's surprising, I think this is what io.Pipe is meant to do. Is it convenient to show me the code that didn't work?

The current code of reading everything and then writing everything probably doesn't exercise as many code paths as interleaving reads and writes would.

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: eh, now that I look, these buffers are not used the logic inside `createHTTPStreams` (copied from the kubelet code IIRC) uses it just to identify the number of streams expected.

only the output streams are used, and if you use io.Copy or a io.Pipe to plumb those together the race detector complains

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: You are right, that's extremely bizarre, I read the createHTTPStreams function, it's super weird.

The streams here seem to be from the spdy library, right? Using io.Copy sounds like it should work to me. I'm actually concerned that the race checker is correct and we may have the same problem in production code. Can you post a stack trace from the race checker?

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: interesting, this works without race, I do not know if there was some change or if I made the wrong test

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]:
```
tress ./remotecommand.test -test.run TestStreamRandomData
5s: 1410 runs so far, 0 failures
10s: 2838 runs so far, 0 failures
15s: 4273 runs so far, 0 failures
20s: 5695 runs so far, 0 failures
25s: 7136 runs so far, 0 failures
30s: 8564 runs so far, 0 failures
35s: 9980 runs so far, 0 failures
40s: 11390 runs so far, 0 failures
45s: 12816 runs so far, 0 failures
```

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:374]: @lavalamp please take a look and get this merged to increase the coverage

**Pull request merged**